### PR TITLE
Handle errors for CSV CRD does not installed

### PR DESCRIPTION
**Pull request by int8dev**
### Motivation
Currently the operator will throw error `"olm","msg":"on listting CSVs","error":"the server could not find the requested resource"` if CSV CRD is not installed.
Handle these errors will make service binding operator work with less dependency(With SBR CRD only).

More detials refer to Issue https://github.com/redhat-developer/service-binding-operator/issues/385

### Changes
Handle errors in `olm.listCSVs()` and also check CSV resources before add watching.

### Testing

The fake client will not return 404 error while listing non-existing resources kind. So test case is not present for this error handling. Maybe we could add cases in e2e test.

### CODE REVIEW

**Comment by sbose78** [comment]
Thank you!

Have you tried out a binding on Kubernetes with no OLM ?

**Comment by int8dev** [comment]
Hi @sbose78 
I have tested with using kubernetes 1.15 cluster with SBR CRD installed only. It works like a charm.

**Comment by sbose78** [comment]
@int8dev Thank you very much!
Do you mind sharing the steps of the scenario that you've tested on kube?

**Comment by int8dev** [comment]
Here is what I have done for testing.
1. Prepare a kubernetes 1.15 cluster with knative installed
2. Configure necessary role and rolebindings and install SBR CRD only. Build service-binding-operator with this commit and deploy to cluster. 
3. Create a backend service using IBM cloud operator following https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_ibmcloud_operator. I choosed a speech-to-text service for testing.
4. Create a simple knative service.
5. Create following ServiceBindingRequest accordingly. 
```apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-speech-to-text
  namespace: default
spec:
  applicationSelector:
    group: serving.knative.dev
    resource: services
    resourceRef: mydemo
    version: v1
  backingServiceSelector:
    group: ibmcloud.ibm.com
    kind: Binding
    resourceRef: my-speech-to-text-binding
    version: v1alpha1
  customEnvVar:
  - name: STX_URL
    value: '{{ index .status.secretName "url" }}'
  - name: STX_APIKEY
    value: '{{ index .status.secretName "apikey" }}'
  detectBindingResources: false
  envVarPrefix: STX
```
6. Check the log of service-binding-operator. Notice there is a warning message complaining that CSV CRD is not installed.
```
{"level":"info","ts":1585231759.7002933,"logger":"reconciler","msg":"Reconciling ServiceBindingRequest...","Request.Namespace":"default","Request.Name":"binding-request-speech-to-text"}
{"level":"info","ts":1585231759.7814934,"logger":"olm","msg":"WARNING: ClusterServiceVersions CRD is not installed"}
{"level":"info","ts":1585231759.8041327,"logger":"retriever","msg":"Looking for spec-descriptors in 'spec'..."}
{"level":"info","ts":1585231759.8041744,"logger":"retriever","msg":"Looking for status-descriptors in 'status'..."}
{"level":"info","ts":1585231759.8191376,"logger":"reconciler","msg":"Starting the bind of application(s) with backing service...","Request.Namespace":"default","Request.Name":"binding-request-speech-to-text","ServiceBindingRequest.Name":"binding-request-speech-to-text"}
{"level":"info","ts":1585231759.819233,"logger":"reconciler.bind","msg":"Binding applications with intermediary secret...","Request.Namespace":"default","Request.Name":"binding-request-speech-to-text","ServiceBindingRequest.Name":"binding-request-speech-to-text"}
{"level":"info","ts":1585231759.8192647,"logger":"reconciler","msg":"Saving data on intermediary secret...","Request.Namespace":"default","Request.Name":"binding-request-speech-to-text","ServiceBindingRequest.Name":"binding-request-speech-to-text"}
{"level":"info","ts":1585231760.0712717,"logger":"reconciler","msg":"All done!","Request.Namespace":"default","Request.Name":"binding-request-speech-to-text","ServiceBindingRequest.Name":"binding-request-speech-to-text"}
```
7. Check the env of ksvc pod
```
root@mydemo-xvh4t-deployment-5b957599b9-sk8zt:/# env|grep STX_
STX_APIKEY=[REDACTED]
STX_BINDING_SECRET_IAM_ROLE_CRN=crn:v1:bluemix:public:iam::::serviceRole:Manager
STX_BINDING_CONFIGMAP_IAM_SERVICEID_CRN=crn:v1:bluemix:public:iam-identity::a/[REDACTED]::serviceid:[REDACTED]
STX_BINDING_CONFIGMAP_IAM_APIKEY_NAME=my-speech-to-text-binding
STX_BINDING_CONFIGMAP_URL=https://api.us-south.speech-to-text.watson.cloud.ibm.com/instances/0f97901f-37a9-4618-b07f-[REDACTED]
STX_BINDING_SECRET_IAM_APIKEY_DESCRIPTION=Auto-generated for key 32d7766f-d7b8-4394-86fc-[REDACTED]
STX_BINDING_SECRET_APIKEY=[REDACTED]
STX_BINDING_SECRET_IAM_SERVICEID_CRN=crn:v1:bluemix:public:iam-identity::a/[REDACTED]::serviceid:ServiceId-426bb652-9860-4d51-a9d3-[REDACTED]
STX_BINDING_SECRET_IAM_APIKEY_NAME=my-speech-to-text-binding
STX_BINDING_CONFIGMAP_IAM_ROLE_CRN=crn:v1:bluemix:public:iam::::serviceRole:Manager
STX_BINDING_SECRET_URL=https://api.us-south.speech-to-text.watson.cloud.ibm.com/instances/0f97901f-37a9-4618-b07f-[REDACTED]
STX_BINDING_CONFIGMAP_IAM_APIKEY_DESCRIPTION=[REDACTED]
STX_URL=https://api.us-south.speech-to-text.watson.cloud.ibm.com/instances/0f97901f-37a9-4618-b07f-[REDACTED]
STX_BINDING_CONFIGMAP_APIKEY=[REDACTED]
```

**Comment by sbose78** [comment]
Thank you!

File: pkg/controller/servicebindingrequest/olm.go [hunk 1]
```diff
@@ -42,7 +42,10 @@ func (o *OLM) listCSVs() ([]unstructured.Unstructured, error) {
 	gvr := olmv1alpha1.SchemeGroupVersion.WithResource(csvResource)
 	resourceClient := o.client.Resource(gvr).Namespace(o.ns)
 	csvs, err := resourceClient.List(metav1.ListOptions{})
-	if err != nil {
+	if err != nil && errors.IsNotFound(err) {
```
Comment (sbose78) [inline pkg/controller/servicebindingrequest/olm.go]: Thank you! do you think a test could be added this?

Comment (int8dev) [reply pkg/controller/servicebindingrequest/olm.go]: I added a test case to mock a not found error like a genuine API server. Please Review. Thanks!

**Comment by sbose78** [comment]
I can confirm that unit tests run locally. I'll override the CI flake and take this in.

@baijum please use https://github.com/redhat-developer/service-binding-operator/pull/386 to debug.

**Comment by sbose78** [comment]
> Configure necessary role and rolebindings and install SBR CRD only.

Hi @int8dev Did you have to modify any rolebindings that were present in `deploy/` ?

**Pull request merged**